types.go

```diff
@@ -0,0 +1,177 @@
+type GeoPath struct {
+	geo.Path
+}
+
+// NewGeoPathFromPointSet create a geo path from points set
+func NewGeoPathFromPointSet(s geo.PointSet) *GeoPath {
+	return &GeoPath{
+		Path: geo.Path{
+			PointSet: s,
+		},
+	}
+}
+
+// Tidy represents a tidy param for osrm5 match request
+type Tidy string
+
+const (
+	TidyTrue  Tidy = "true"
+	TidyFalse Tidy = "false"
+)
+
+// String returns Tidy as a string
+func (t Tidy) String() string {
+	return string(t)
+}
+
+// Annotations represents a annotations param for osrm5 request
+type Annotations string
+
+const (
+	AnnotationsTrue  Annotations = "true"
+	AnnotationsFalse Annotations = "false"
+)
+
+// String returns Annotations as a string
+func (a Annotations) String() string {
+	return string(a)
+}
+
+// Steps represents a steps param for osrm5 request
+type Steps string
+
+const (
+	StepsTrue  Steps = "true"
+	StepsFalse Steps = "false"
+)
+
+// String returns Steps as a string
+func (s Steps) String() string {
+	return string(s)
+}
+
+// Gaps represents a gaps param for osrm5 match request
+type Gaps string
+
+const (
+	GapsSplit  Gaps = "split"
+	GapsIgnore Gaps = "ignore"
+)
+
+// String returns Gaps as a string
+func (g Gaps) String() string {
+	return string(g)
+}
+
+// Geometries represents a geometries param for osrm5
+type Geometries string
+
+const (
+	GeometriesPolyline6 Geometries = "polyline6"
+	GeometriesGeojson   Geometries = "geojson"
+)
+
+// String returns Geometries as a string
+func (g Geometries) String() string {
+	return string(g)
+}
+
+// Overview represents level of overview of geometry in a response
+type Overview string
+
+// Available overview values
+const (
+	OverviewSimplified Overview = "simplified"
+	OverviewFull       Overview = "full"
+	OverviewFalse      Overview = "false"
+)
+
+// String returns Overview as a string
+func (o Overview) String() string {
+	return string(o)
+}
+
+// Request contains parameters for OSRM query
+type Request struct {
+	Profile string
+	GeoPath GeoPath
+
+	service string
+	options Options
+}
+
+// Response contains properties from OSRM query
+type Response interface {
+	Error
+}
+
+// URL generates a url for OSRM request
+func (r *Request) URL(serverURL string) (string, error) {
+	if r.service == "" {
+		return "", ErrEmptyServiceName
+	}
+	if r.Profile == "" {
+		return "", ErrEmptyProfileName
+	}
+	if r.GeoPath.Length() == 0 {
+		return "", ErrNoCoordinates
+	}
+	// http://{server}/{service}/{version}/{profile}/{coordinates}[.{format}]?option=value&option=value
+	url := strings.Join([]string{
+		serverURL, // server
+		r.service, // service
+		version,   // version
+		r.Profile, // profile
+		"polyline(" + r.GeoPath.Polyline() + ")", // coordinates
+	}, "/")
+	if len(r.options) > 0 {
+		url += "?" + r.options.Encode() // options
+	}
+	return url, nil
+}
+
+// Polyline generates a polyline in Google format
+// It uses default factor because of OSRM5 doesn't support polyline6 as coordinates
+func (g *GeoPath) Polyline() string {
```

Comment:
Could you move functions near a definition?

Reply:
updated